Eastern-Research-Group/expert-query
# Feature/319 improve retry fetch

## Related Issues:
* [EQ-319](https://jira.epa.gov/browse/EQ-319)

## Main Changes:
* Removed an extra slash from the ordspub service calls.
* Improved the retry axios call logic.
  * Updated the retry logic to better handle bad status codes (i.e., 502). Before, a 502 status code would cause the etl to just bomb out without retrying. 
  * Updated all of our retry axios calls to use the same `fetchRetry` function. Before, we had different retry logic in some places.

## Steps To Test:
1. Verify etl is pointing to local db
2. In `tableConfig.json` update the `createQuery` value for the actions profile, such that `watertype`, `watersize`, and `watersizeunits` allow nulls. 
    * It looks like we are in a weird spot where attains has updated some of the materialized views, but our S3 bucket hasn't been refreshed. So the ordspub services are allowing nulls in those columns above. This is a good reason to try moving away from ordspub services, like Paul was suggesting.
    * I'll make a separate commit to `tableConfig.json` once we get a full list of schema changes.
3. Run the etl
4. Verify database part succeeded
5. Verify glossary part succeeded
6. Verify domain_values part succeeded
7. Check the console. If it has one or more entries that say `Non-200 response returned from profile_stats service, retrying`, then skip the remaining items
8. To force retries, on line 640 of `etl\app\server\database.js`, add a low timeout the `callOptions` parameter (example below). 

```
  const res = await fetchRetry({
    url,
    s3Config,
    serviceName: 'profile_stats',
    retryInterval: s3Config.config.retryIntervalProfileStatsSeconds,
    callOptions: {
      timeout: 5000,
      headers: { 'API-key': process.env.MV_API_KEY },
      httpsAgent: new https.Agent({
        // TODO - Remove this when ordspub supports OpenSSL 3.0
        // This is needed to allow node 18 to talk with ordspub, which does
        //   not support OpenSSL 3.0
        secureOptions: crypto.constants.SSL_OP_LEGACY_SERVER_CONNECT,
      }),
    },
  });
```
